src/ucx_rma_plugin.c

```diff
@@ -0,0 +1,1075 @@
+    nccl_ucx_rma_init_ucp(dev, &workers[dev].ctx);
+    nccl_ucx_rma_init_worker(workers[dev].ctx, &workers[dev].worker);
+    workers->count = 0;
+    workers->eps   = NULL;
+  }
+
+  comm_ctx->ctx    = workers[dev].ctx;
+  comm_ctx->worker = workers[dev].worker;
+  comm_ctx->id     = workers[dev].count;
+  workers[dev].count++;
+#else
+  nccl_ucx_rma_init_ucp(dev, &comm_ctx->ctx);
+  nccl_ucx_rma_init_worker(comm_ctx->ctx, &comm_ctx->worker);
+#endif
+  pthread_mutex_unlock(&nccl_ucx_rma_lock);
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_ucx_rma_send_worker_address(ucp_worker_h worker, int fd)
+{
+  ucp_worker_attr_t attr;
+
+  attr.field_mask    = UCP_WORKER_ATTR_FIELD_ADDRESS |
+                       UCP_WORKER_ATTR_FIELD_ADDRESS_FLAGS;
+  attr.address_flags = UCP_WORKER_ADDRESS_FLAG_NET_ONLY;
+
+  UCXCHECK(ucp_worker_query(worker, &attr));
+  NCCLCHECK(socketSend(fd, &attr.address_length, sizeof(attr.address_length)));
+  NCCLCHECK(socketSend(fd, attr.address, attr.address_length));
+
+  free(attr.address);
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_ucx_free_worker(ucp_worker_h worker)
+{
+  int i;
+  int dummy;
+  struct ep_list *ep, *cur;
+
+  pthread_mutex_lock(&nccl_ucx_rma_lock);
+  for(i = 0; i < ncclNIbDevs; i++) {
+    if (worker == workers[i].worker) {
+      workers[i].count--;
+      if (workers[i].count == 0) {
+        ep = workers[i].eps;
+        while(ep) {
+          cur = ep;
+          NCCLCHECK(socketReceive(ep->fd, &dummy, sizeof(int)));
+          ep = ep->next;
+          close(cur->fd);
+          free(cur);
+        }
+        ucp_worker_destroy(workers[i].worker);
+        ucp_cleanup(workers[i].ctx);
+        INFO(NCCL_NET, "worker destroy");
+        workers[i].eps    = NULL;
+        workers[i].worker = NULL;
+        workers[i].ctx    = NULL;
+      }
+      break;
+    }
+  }
+  pthread_mutex_unlock(&nccl_ucx_rma_lock);
+
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_ucx_add_ep(ucp_worker_h worker, int fd)
+{
+  ncclResult_t status = ncclSuccess;
+  int i;
+
+  for(i = 0; i < ncclNIbDevs; i++) {
+    if (worker == workers[i].worker) {
+      struct ep_list *new_ep = (struct ep_list*)malloc(sizeof(struct ep_list));
+
+      if (new_ep == NULL) {
+        status = ncclSystemError;
+        break;
+      }
+
+      new_ep->fd   = fd;
+      new_ep->next = workers[i].eps;
+      workers[i].eps = new_ep;
+      break;
+    }
+  }
+
+  return status;
+}
+
+ncclResult_t nccl_ucx_rma_init(ncclDebugLogger_t logFunction)
+{
+  char *config_env;
+  if (ncclParamUCXRMADisable()) return ncclInternalError;
+  NCCLCHECK(nccl_p2p_ib_init(&ncclNIbDevs, ncclIbDevs, if_name, &nccl_ucx_if_addr,
+                          NULL, logFunction));
+
+  if (strlen(nccl_ucx_rma_tls) == 0) {
+    config_env = getenv("NCCL_UCX_TLS");
+    if (config_env != NULL) {
+      snprintf(nccl_ucx_rma_tls, 32, "%s", config_env);
+    } else {
+      snprintf(nccl_ucx_rma_tls, 32, "%s", "ib");
```

Comment:
dont we need any cuda transport?

Reply:
I don't think we need cuda transport for RMA plugin mainly because we rely on gdr as nccl ib does. Also AFAIK ucp_put doesn't support cuda

Reply:
BTW, do we need to save this configuration in global static variables? can we directly use them in nccl_ucx_rma_init_ucp().   ucp_config_modify() if user not set